### PR TITLE
Acquire read lock when getting file or directory info

### PR DESCRIPTION
**Pull request by icewind1991**
Prevents cache data being changed while a file is being written or a folder is being renamed.

For #13391

cc @DeepDiver1975 @PVince81

### CODE REVIEW

File: lib/private/files/view.php [hunk 1]
```diff
@@ -674,6 +674,7 @@ public function rename($path1, $path2) {
 					}
 				} elseif ($storage1 == $storage2) {
 					if ($storage1) {
+						sleep(20);
```
Comment (MorrisJobke) [inline lib/private/files/view.php]: I hope this is debug stuff. Or is this for some future performance improvements? ;)

Comment (LukasReschke) [inline lib/private/files/view.php]: No. This reduces the risk of race conditions :speak_no_evil: :laughing:

Comment (icewind1991) [inline lib/private/files/view.php]: And also reduces server load by giving the cpu some time to catch up

Comment (LukasReschke) [inline lib/private/files/view.php]: All in all, we need to add more of such statements.

File: lib/private/connector/sabre/objecttree.php [hunk 2]
```diff
@@ -150,6 +150,8 @@ public function getNodeForPath($path) {
 				throw new \Sabre\DAV\Exception\ServiceUnavailable('Storage not available');
 			} catch (StorageInvalidException $e) {
 				throw new \Sabre\DAV\Exception\NotFound('Storage ' . $path . ' is invalid');
+			} catch (LockedException $e) {
+				throw new \Sabre\DAV\Exception\Locked();
```
Comment (DeepDiver1975) [inline lib/private/connector/sabre/objecttree.php]: This reminds me that we should add locking info to the webdav interface.

@icewind1991 @PVince81 I guess this is off scope for 8.1 ?

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

File: lib/private/connector/sabre/objecttree.php (hunk 2 above)
Comment (PVince81) [inline lib/private/connector/sabre/objecttree.php]: Raised https://github.com/owncloud/core/issues/16968

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Cannot test properly, requires https://github.com/owncloud/core/issues/16959 (https://github.com/owncloud/core/pull/16955) to fix upload

**Comment by PVince81** [comment]
Ok I managed to upload the files using another way.

Renaming the folder while the sync client is running still lead to loss of its contents (after 5 test runs).

I'll do another run where I merge https://github.com/owncloud/core/pull/16729 into this one to see if it improves it.

**Comment by PVince81** [comment]
Ok sorry, wrong branch... will retest again.

**Comment by PVince81** [comment]
Data loss still occurring (renamed folder empty), with the following in the log:
```
{"reqId":"CTjGU7ARO9ANbFxO0f5O","remoteAddr":"127.0.0.1","app":"core","message":"Exception while scanning file \"files\/re\": An exception occurred while executing 'INSERT INTO `oc_filecache` (`mimepart`,`mimetype`,`mtime`,`size`,`etag`,`storage_mtime`,`permissions`,`parent`,`path_hash`,`path`,`name`,`storage`) SELECT ?,?,?,?,?,?,?,?,?,?,?,? FROM `oc_filecache` WHERE `storage` = ? AND `path_hash` = ? HAVING COUNT(*) = 0' with params [\"1\", \"2\", 1434466525, -1, \"55803938d795e\", 1434466525, 31, \"2\", \"ade12aa9a2fcb78d16c649c813e122c6\", \"files\\\/re\", \"re\", \"1\", \"1\", \"ade12aa9a2fcb78d16c649c813e122c6\"]:\n\nSQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction","level":0,"time":"2015-06-16T14:57:33+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/"}
```

Next up: same test steps with https://github.com/owncloud/core/pull/16729 merged

**Comment by PVince81** [comment]
Hmm, too bad. Even with this PR and https://github.com/owncloud/core/pull/16729 together, the contents of the renamed folder appeared empty again in my test case.

This appeared in the log:
```
{"reqId":"J79porgAwfcvfd0WcVdh","remoteAddr":"127.0.0.1","app":"index","message":"Exception: {\"Exception\":\"Doctrine\\\\DBAL\\\\Exception\\\\DriverException\",\"Message\":\"An exception occurred while executing 'DELETE FROM `oc_filecache` WHERE `parent` = ?' with params [288]:\\n\\nSQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(116): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Statement.php(172): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'DELETE FROM `oc...', Array)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/db\\\/statementwrapper.php(73): Doctrine\\\\DBAL\\\\Statement->execute(Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/db.php(147): OC_DB_StatementWrapper->execute(Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/cache.php(500): OC_DB::executeAudited('DELETE FROM `*P...', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/cache.php(471): OC\\\\Files\\\\Cache\\\\Cache->removeChildren(Array)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(200): OC\\\\Files\\\\Cache\\\\Cache->remove('files\\\/re')\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(352): OC\\\\Files\\\\Cache\\\\Scanner->removeFromCache('files\\\/re')\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(257): OC\\\\Files\\\\Cache\\\\Scanner->scanChildren('files', false, 3, Array)\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/watcher.php(91): OC\\\\Files\\\\Cache\\\\Scanner->scan('files', false)\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1173): OC\\\\Files\\\\Cache\\\\Watcher->checkUpdate('files', Array)\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/filesystem.php(861): OC\\\\Files\\\\View->getFileInfo('\\\/', false)\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/helper.php(968): OC\\\\Files\\\\Filesystem::getFileInfo('\\\/', false)\\n#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/lib\\\/helper.php(45): OC_Helper::getStorageInfo('\\\/')\\n#14 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/ajax\\\/getstoragestats.php(37): OCA\\\\Files\\\\Helper::buildFileStorageStatistics('\\\/')\\n#15 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/route.php(153) : runtime-created function(1): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#16 [internal function]: __lambda_func(Array)\\n#17 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/route\\\/router.php(274): call_user_func('\\\\x00lambda_4', Array)\\n#18 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/base.php(867): OC\\\\Route\\\\Router->match('\\\/apps\\\/files\\\/aja...')\\n#19 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/index.php(40): OC::handleRequest()\\n#20 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":115}","level":4,"time":"2015-06-16T15:18:14+00:00","method":"GET","url":"\/owncloud\/index.php\/apps\/files\/ajax\/getstoragestats.php?dir=%2F"}
```

This seems similar to what I observed in the original issue (a failed DELETE followed by a successful one) https://github.com/owncloud/core/issues/13391

@icewind1991 let me know if you need more details to reproduce this.
Most important is having at least 100 files in the folder and the sleep(1) in place, see https://github.com/owncloud/core/issues/13391#issuecomment-105525369

**Comment by PVince81** [comment]
I'd say the key is having the rename operation slow enough so that the running sync client can do update checks several times within that time. One of them is likely to cause the data loss. Other cases will cause redownloads. In my env, there chance is 70% of data loss.

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by icewind1991** [comment]
switched to an approach where when getting the lock fails it uses the old cache data since it looks like there are to many code path relying on being able to get the cache data during the operation

**Comment by PVince81** [comment]
Have you managed to reproduce the issue this time ?

**Comment by icewind1991** [comment]
No, what are the steps you're using to reproduce?

**Comment by PVince81** [comment]
Come on, I already mentioned them so many times every time I tell you that it did not fix the issue. Here https://github.com/owncloud/core/issues/13391#issuecomment-105525369

**Comment by PVince81** [comment]
- [ ] BUG: cannot create directory (master is fine with the same steps)
1. Reset ownCloud
2. Clear redis cache with "flushall"
3. Setup ownCloud
4. Enable experimental file locking 
5. Setup sync client for admin user (make sure it's from scratch, empty sync data dir)
6. Run sync, it downloads "welcome.txt"
7. Create a folder "food"
8. Wait for sync
```
{"reqId":"nPBfl1MbxpzqmPh+mTub","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"food\\\" is locked\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(1177): OC\\\\Connector\\\\Sabre\\\\Directory->createDirectory('food')\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(583): Sabre\\\\DAV\\\\Server->createCollection('food', Array, Array)\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMkcol(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:MKCOL', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(84): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#8 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/connector\\\/sabre\\\/directory.php\",\"Line\":147}","level":4,"time":"2015-06-17T13:14:36+00:00","method":"MKCOL","url":"\/owncloud\/remote.php\/webdav\/food"}
```

**Comment by PVince81** [comment]
Still does not fix https://github.com/owncloud/core/issues/13391#issuecomment-105525369.
Reproduced after 3 tries.
This appeared in the log:
```
{"reqId":"vO5YWrSJQ48S7tx5HqNr","remoteAddr":"127.0.0.1","app":"webdav","message":"Exception: {\"Message\":\"An exception occurred while executing 'DELETE FROM `oc_filecache` WHERE `parent` = ?' with params [9]:\\n\\nSQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/DBALException.php(116): Doctrine\\\\DBAL\\\\Driver\\\\AbstractMySQLDriver->convertException('An exception oc...', Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Statement.php(172): Doctrine\\\\DBAL\\\\DBALException::driverExceptionDuringQuery(Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOMySql\\\\Driver), Object(Doctrine\\\\DBAL\\\\Driver\\\\PDOException), 'DELETE FROM `oc...', Array)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/db\\\/statementwrapper.php(73): Doctrine\\\\DBAL\\\\Statement->execute(Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/db.php(147): OC_DB_StatementWrapper->execute(Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/cache.php(500): OC_DB::executeAudited('DELETE FROM `*P...', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/cache.php(471): OC\\\\Files\\\\Cache\\\\Cache->removeChildren(Array)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(200): OC\\\\Files\\\\Cache\\\\Cache->remove('files\\\/re2')\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(352): OC\\\\Files\\\\Cache\\\\Scanner->removeFromCache('files\\\/re2')\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/cache\\\/scanner.php(257): OC\\\\Files\\\\Cache\\\\Scanner->scanChildren('files', false, 3, Array)\\n#9 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/files\\\/view.php(1257): OC\\\\Files\\\\Cache\\\\Scanner->scan('files', false)\\n#10 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/private\\\/connector\\\/sabre\\\/directory.php(196): OC\\\\Files\\\\View->getDirectoryContent('\\\/')\\n#11 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(195): OC\\\\Connector\\\\Sabre\\\\Directory->getChildren()\\n#12 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(878): Sabre\\\\DAV\\\\Tree->getChildren('')\\n#13 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(928): Sabre\\\\DAV\\\\Server->addPathNodesRecursively(Array, Object(Sabre\\\\DAV\\\\PropFind))\\n#14 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(327): Sabre\\\\DAV\\\\Server->getPropertiesForPath('', Array, 1)\\n#15 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPropfind(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#16 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#17 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:PROPFIND', Array)\\n#18 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#19 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#20 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(84): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#21 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/3rdparty\\\/doctrine\\\/dbal\\\/lib\\\/Doctrine\\\/DBAL\\\/Driver\\\/AbstractMySQLDriver.php\",\"Line\":115}","level":4,"time":"2015-06-17T13:21:32+00:00","method":"PROPFIND","url":"\/owncloud\/remote.php\/webdav\/"}
```
It looks like somehow the scanner still detected the missing file and tried and delete the entries.

**Comment by icewind1991** [comment]
From what I can tell this doesn't fix the issue with your reproduction steps because atm only a read lock is hold around the cache rename, I'll experiment further with it

**Comment by icewind1991** [comment]
I'm a bit stuck here, the cache needs to be locked to only have one process changing the cache for a folder at the time separate from the lock we have in place to prevent the underlying files being changed while updating the cache.

And those locks need to work together to prevent any race conditions.

**Comment by PVince81** [comment]
Do we need a new lock type for the cache maybe ?

**Comment by PVince81** [comment]
Back then I tried to do that with DB transaction here https://github.com/owncloud/core/pull/16576 (but I was told to be careful with transactions)

**Comment by icewind1991** [comment]
That was I was thinking of but I'm not sure what the correct way is to handle those locks correctly.

We can't just acquire a cache lock once we start doing a scan/move/etc since then a different process might have already updated the cache between the moment we're done with updating the file but before the cache update has started.

Transactions wouldn't fully work for the same reason

**Comment by icewind1991** [comment]
Maybe it will work if we have an atomic way to go from a write lock on a file to both a read lock on the file and a write lock on the cache for the file

**Comment by icewind1991** [comment]
Trying to get a write lock on the cache should not only fail when the cache is already locked but also when the file for the cache is write locked

**Comment by PVince81** [comment]
Not sure I understand the problem properly.
The problem I saw with the DB transaction was simply that the transaction doesn't wrap the actual on-disk file-operation. The DB transaction is done after the disk part has been updated yet.

Is that what you meant ?

Originally when thinking about file locking I was hoping for a big single transaction that wraps everything. For example the transaction would start in ajax/rename.php before the rename and ends at the end of ajax/rename.php when all calls returned. But now the current locking is locking/unlocking several times within a single file operation, due to it being on View level.

**Comment by PVince81** [comment]
Maybe you could explain the concurrency using process steps.

Say we have process A and process B. A is for folder rename and B for the PROPFIND.
1. A: $view->rename()
1. A: acquire cache lock on "folder" and "folder_rename"
1. A: acquire share lock on "folder" and "folder_rename"
1. A: run pre-hook
1. A: change locks to exclusive on "folder" and "folder_rename"
1. A: do on-disk rename
1. A: update database entries for the rename
1. B: PROPFIND
1. B: calls `hasUpdated`
1. B: acquire lock on cache folder "folder_rename" which was found on disk
1. B: lock fail, skip folder
1. B: return PROPFIND result
1. A: continue update database entries for the rename
1. A: finish update database entries for the rename
1. A: change locks to back to shared on "folder" and "folder_rename" (still todo in another ticket)
1. A: run post-hooks
1. A: release cache hook

Ok so at step 10 it is not clear on which folder to set the cache lock ?

**Comment by icewind1991** [comment]
Currently we have

1. get read lock
2. pre-hook
3. change to write lock
4. do rename/write/whatever
5. change to read lock
6. update cache
7. release lock (post hooks can be inserted before this without messing with this issue I think)

We need to prevent any other process from trying to change (maybe also read) the cache at any point from before step 4 until after step 6.

Currently we block the scanner from steps 3 to 5 which can be extended to also include other cache updating operations (delete, move, copy).
The problem is that after step 5 we no longer have this protection since we need to make the file readable for our own scanner and once we do that scanners from other processes have read access to.

We can only block cache reads during step 6, any other time it will break stuff that relies on reading the cache, if we go for only blocking writes during the cache we can use a wider lock but then we need to either cache to an exclusive cache lock around step 6 or make sure every cache operation is atomic.

My current, untested, idea would be something like

1. get read lock
2. pre-hook
3. change to write lock
4. do rename/write/whatever
5. change to read lock + cache write lock (one atomic change, otherwise :boom: )
6. update cache
7. release cache write lock
8. post hook
9. release read lock

And try to acquire a cache read lock on `getFileInfo`/`getFolderContent`

**Comment by PVince81** [comment]
Argh, and I see we also still have the propagator (size, etag) which would affect parent folders as well, additionally to the regular cache updates, see https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L703.

To make it atomic, we could introduce a new type "SHARED_PLUS_CACHE" which could be set using changeLock. But this is assuming that we want to use the same paths like the other locks.

I guess we don't want/need to set the cache lock on parent folders of the action ? Note that these might get updated (size, etag).

**Comment by PVince81** [comment]
Your proposal sounds good.

**Comment by PVince81** [comment]
(I'm still afraid that this is getting more complex over time and I wonder if we will ever manage to fix this properly without major architectural changes)

**Comment by icewind1991** [comment]
> To make it atomic, we could introduce a new type "SHARED_PLUS_CACHE" which could be set using changeLock. But this is assuming that we want to use the same paths like the other locks.

I don't think we can have both the file and cache lock on the same (memcache) key since the shared lock needs to be incrementable.

A different way might be to have the cache update within the exclusive file lock, some logic would need to be added so that it doesn't try to get the read lock in the scanner when the scanner is called from the cache update (but only from scans such as triggered by `getFileInfo`).

That way we have the exclusive file lock for any cache update which is triggered from a view operation to prevent concurrent changes and a read lock when doing an externally triggered scan (2 concurrent scans should not lead to data corruption since the files are read locked)

**Comment by PVince81** [comment]
Yes, also had a similar idea: the scan commands running within the file ops get "special rights" to bypass the lock.

Well it's actually more a cache update than scan. But in some cases we resort to scanFile() just to get the info from disk. Hopefully we can change that some day.

**Comment by icewind1991** [comment]
> BUG: cannot create directory (master is fine with the same steps)

Should be fixed now

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Fixes https://github.com/owncloud/core/issues/16998

**Comment by PVince81** [comment]
@icewind1991 I'm still seeing https://github.com/owncloud/core/pull/16963#issuecomment-112795034, cannot create directory in the root with the sync client. "Locked"

**Comment by PVince81** [comment]
Wait, I mismanipulated the branches. Trying again.

**Comment by PVince81** [comment]
Works now :+1:

**Comment by nickvergessen** [comment]
PR fixes the failing smashbox tests

File: lib/private/files/view.php [hunk 3]
```diff
@@ -1805,6 +1826,8 @@ public function lockFile($path, $type) {
 	 * @return bool False if the path is excluded from locking, true otherwise
 	 */
 	public function unlockFile($path, $type) {
+		$path = '/' . trim($path, '/');
```
Comment (nickvergessen) [inline lib/private/files/view.php]: shouldn't this be covered by `Filesystem::normalizePath($absolutePath)` nowerdays?

Comment (PVince81) [inline lib/private/files/view.php]: True. @icewind1991 probably forgot to remove it after rebasing ? (normalizePath wasn't there originally)

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
I will test all smashbox tests agains this. So far it looks promising

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

**Comment by nickvergessen** [comment]
Test | Result
---- | ------
test_reshareDir -a | :white_check_mark: 
test_shareDir -a | :white_check_mark: 
test_shareFile -a | :white_check_mark: 
test_shareGroup -a | :white_check_mark: 
test_uploadFiles -a | :white_check_mark: 
test_basicSync -a | :white_check_mark: 

sharePermissions fails, so I split it up for details:

Test | Result
---- | ------
test_sharePermissions -t=0 (ALL) | :x: move_in, move_in_overwrite and move_out fail
test_sharePermissions -t=1 (READ) | :white_check_mark: 
test_sharePermissions -t=2 (READ + CREATE) | :x: move_in
test_sharePermissions -t=3 (READ + UDPATE) | :white_check_mark: 
test_sharePermissions -t=4 (READ + DELETE) | :x: move_out
test_sharePermissions -t=5 (READ + CREATE + UPDATE) | :x: move_in
test_sharePermissions -t=6 (READ + CREATE + DELETE) | :x: move_in, move_in_overwrite and move_out
test_sharePermissions -t=7 (READ + UPDATE + DELETE) | :x: move_out

Note: move_in_overwrite deletes + creates thats why READ+UPDATE is okay.

**Comment by nickvergessen** [comment]
So :+1: since it fixes 4 out of the 5 failing tests.
sharePermissions will be dealt with in another ticket

**Comment by nickvergessen** [comment]
New issue created https://github.com/owncloud/core/issues/17029

**Pull request merged**